### ISSUE TITLE
Zod preprocess does not generate error

### ISSUE THREAD

**Issue by dhulme** [react-hook-form/resolvers#635]
**Describe the bug**
Using z.preprocess() instead of z.coerce() in a validation schema does not generate an error. When using zod directly, both of these approaches generate the same error.

**To Reproduce**

1. On codesandbox link, click submit button
2. 'Age coerce' input generates an error, but 'Age preprocess' does not
3. Note how the errors produced from zod directly are identical in the console

**Codesandbox link (Required)**
https://codesandbox.io/s/react-hook-form-zod-preprocess-mn8sy4?file=/src/index.js

**Expected behavior**
An error should be generated

**Screenshots**

**Desktop (please complete the following information):**

- OS: Windows 10
- Browser: Chrome
- Version : 117

**Additional context**

**Comment by yukukotani** [comment]
I have the same issue here.

Interestingly, it works if `agePreprocess` condition is the first property of the schema object.

```diff
 const SignupSchema = z.object({
-  ageCoerce,
-  agePreprocess,
+  agePreprocess,
+  ageCoerce,
 });
```

**Comment by dhulme** [comment]
That is interesting. I can confirm the same behaviour.

**Comment by stackline** [comment]
@dhulme

I have the same issue.
I downgraded the zod version from 3.22.3 to 3.21.4 and the issue was resolved, so I would like to share it.

**Codesandbox link (Changed zod version from 3.22.3 to 3.21.4)**
https://codesandbox.io/s/react-hook-form-zod-preprocess-forked-cqm66q

**Comment by Austin-Love** [comment]
We have found the same behavior with the fields being order-depdendent. We replaced our preprocessors with transforms, and they work similarly to how the preprocessors worked in 3.21.4. Downgrading is not an option for us due to a CVE in 3.21.4 our cyber team flagged.

**Comment by stackline** [comment]
@dhulme 

The issue commented on https://github.com/react-hook-form/resolvers/issues/635#issuecomment-1778665491 has been resolved by the following pr fix for zod, so I would like to share it.

- https://github.com/colinhacks/zod/pull/2912

And this fix is included in v3.23.0 of zod.

- https://github.com/colinhacks/zod/releases/tag/v3.23.0

**Comment by dhulme** [comment]
@stackline Confirmed as fixed, thanks!
